lib/graphs/digraph.nit

```diff
@@ -0,0 +1,606 @@
+	end
+
+	# The recursive part of Tarjan's algorithm
+	private fun tarjan(u: V)
+	do
+		tarjan_vertex_to_index[u] = tarjan_index
+		tarjan_ancestor[u] = tarjan_index
+		tarjan_index += 1
+		tarjan_stack.add(u)
+		tarjan_in_stack[u] = true
+		for v in successors(u) do
+			if not tarjan_vertex_to_index.keys.has(v) then
+				tarjan(v)
+				tarjan_ancestor[u] = tarjan_ancestor[u].min(tarjan_ancestor[v])
+			else if tarjan_in_stack[v] then
+				tarjan_ancestor[u] = tarjan_ancestor[u].min(tarjan_vertex_to_index[v])
+			end
+		end
+		if tarjan_vertex_to_index[u] == tarjan_ancestor[u] then
+			var v
+			loop
+				v = tarjan_stack.take
+				tarjan_in_stack[v] = false
+				sccs.union(u, v)
+				if u == v then break
+			end
+		end
+	end
+
+	# The strongly connected components computed in Tarjan's algorithm
+	private var sccs: DisjointSet[V] is noinit
+	# An index used for Tarjan's algorithm
+	private var tarjan_index: Int is noinit
+	# A stack used for Tarjan's algorithm
+	private var tarjan_stack: Queue[V] is noinit
+	# A map associating with each vertex its index
+	private var tarjan_vertex_to_index: HashMap[V, Int] is noinit
+	# A map associating with each vertex its ancestor in Tarjan's algorithm
+	private var tarjan_ancestor: HashMap[V, Int] is noinit
+	# True if and only if the vertex is in the stack
+	private var tarjan_in_stack: HashMap[V, Bool] is noinit
+end
+
+# A directed graph represented by hash maps
+class HashMapDigraph[V: Object]
+	super AbstractDigraph[V]
+
+	# Attributes
+	private var pred_map = new HashMap[V, Array[V]]
+	private var succ_map = new HashMap[V, Array[V]]
+
+	redef fun add_vertex(u: V)
+	do
+		if has_vertex(u) then
+			return
+		else
+			pred_map[u] = new Array[V]
+			succ_map[u] = new Array[V]
+			num_vertices += 1
+		end
+	end
+
+	redef fun has_vertex(u: V): Bool
+	do
+		return pred_map.keys.has(u)
+	end
+
+	redef fun remove_vertex(u: V)
+	do
+		if has_vertex(u) then
+			for v in successors(u) do
+				remove_arc(u, v)
+			end
+			for v in predecessors(u) do
+				remove_arc(v, u)
+			end
+			pred_map.keys.remove(u)
+			succ_map.keys.remove(u)
+			num_vertices -= 1
+		end
+	end
+
+	redef fun add_arc(u, v: V)
+	do
+		if not has_vertex(u) then add_vertex(u)
+		if not has_vertex(v) then add_vertex(v)
+		if not has_arc(u, v) then
+			succ_map[u].add(v)
+			pred_map[v].add(u)
+			num_arcs += 1
+		end
+	end
+
+	redef fun has_arc(u, v: V): Bool
+	do
+		return succ_map[u].has(v)
+	end
+
+	redef fun remove_arc(u: V, v: V)
+	do
+		if has_vertex(u) and has_vertex(v) and has_arc(u, v) then
+			succ_map[u].remove(v)
+			pred_map[v].remove(u)
+			num_arcs -= 1
+		end
+	end
+
+	redef fun predecessors(u: V): Array[V]
+	do
+		if pred_map.keys.has(u) then
+			return pred_map[u]
+		else
+			return new Array[V]
+		end
+	end
+
+	redef fun successors(u: V): Array[V]
+	do
+		if succ_map.keys.has(u) then
+			return succ_map[u]
+		else
+			return new Array[V]
+		end
+	end
+
+	redef fun vertices: RemovableCollection[V]
+	do
+		return succ_map.keys
```

Comment:
because you return a Removable collection, people may invalidate the data structure because modification on `succ_map.keys` will not be applied on `pred_map` or `num_vertices`.